### ISSUE TITLE
Investigate `weaktest.ml` failure

### ISSUE THREAD

**Issue by Sudha247** [ocamllabs/ocaml-multicore#547]
`weaktest.ml` is currently disabled in the testsuite. The disabled document mentions this is due to incompatibility with `Obj.truncate`. The comment is now outdated because `Obj.truncate` has since been deprecated in trunk and its usage in `Weak` module has been removed in https://github.com/ocaml/ocaml/pull/2279. So, essentially `Weak` module does not use `Obj.truncate` anymore. Although, the test still fails, which may need investigation.

**Comment by abbysmal** [comment]
After looking into it with @Sudha247, I think I found something
My understanding of our ephemerons implementation is not the best, however I believe I did spot the problematic behavior.

This boils down to `weaktest.ml`, Weak Hashsets and our implementation of `caml_ephe_blit`.

Looking at trunk, on blit, if the GC phase is `Phase_clean`, the runtime goes through `caml_ephe_clean_partial`.
https://github.com/ocaml/ocaml/blob/trunk/runtime/weak.c#L548

Our behavior is such:
https://github.com/ocaml-multicore/ocaml-multicore/blob/4.12%2Bdomains%2Beffects/runtime/weak.c#L368

What I also gather is that `caml_ephe_clean_partial` is doing something very similar to our marking code, except it is not marking, but rather, scanning the array to individually set unmarked values to `caml_ephe_none`.
https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/weak.h#L164

What this means is that on blit, during Phase_clean, the ephemeron data is released.
This is not the case the case on our implementation, and it leads to the test failures we can witness, where the elements keeps pilling up and the bucket keep increasing in size (leading to `weaktest.ml` to fail, and to a tremendous slowdown.)

This would explains the logs we see for weaktest, comparing the `trunks` numbers and the ones we see on our branch, see https://gist.github.com/Sudha247/d0df0d4aa032029d61100e52a816b4f7 for reference.
